queuebacca-core/src/main/java/io/axonif/queuebacca/subscribing/MessageConsumerSelector.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 The Queuebacca Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.axonif.queuebacca.subscribing;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import io.axonif.queuebacca.Message;
+import io.axonif.queuebacca.MessageConsumer;
+
+class MessageConsumerSelector<M extends Message> {
+
+    private final NavigableMap<Integer, MessageConsumer<M>> consumers;
+
+    private MessageConsumerSelector(NavigableMap<Integer, MessageConsumer<M>> consumers) {
+        this.consumers = consumers;
+    }
+
+    static <M extends Message> Builder<M> builder(MessageConsumer<M> firstMessageConsumer) {
+        requireNonNull(firstMessageConsumer);
+
+        NavigableMap<Integer, MessageConsumer<M>> consumers = new TreeMap<>();
+        consumers.put(0, firstMessageConsumer);
+        return new Builder<>(consumers);
+    }
+
+    static <M extends Message> Builder<M> builder(MessageConsumerSelector<M> existingSelector) {
+        requireNonNull(existingSelector);
+        return new Builder<>(new TreeMap<>(existingSelector.consumers));
+    }
+
+    MessageConsumer<M> select(int readCount) {
+        return consumers.floorEntry(readCount).getValue();
+    }
+
+    int currentReadCountThreshold(int readCount) {
+        return consumers.subMap(0, readCount).keySet().stream()
+                .mapToInt(v -> v)
+                .sum();
```

Comment:
I haven't read through everything yet, but why a sum? I would think this should be the max read count key which is mapped to a consumer.

Reply:
It's additive. So if you had 2 fallback consumers, you'd register them for the number of retries between the one before it, e.g. primary -> fallback 1 after 15 reads -> fallback 2 after 5 reads = 20 reads total before we hit fallback 2.